controllers/release/release_adapter.go

```diff
@@ -0,0 +1,306 @@
+		Namespace: a.release.Namespace,
+		Name:      a.release.Spec.ReleaseLink,
+	}, releaseLink)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return releaseLink, nil
+}
+
+// getReleasePipelineRun returns the PipelineRun referenced by the Release being processed. If the PipelineRun is not
+// found or the List operation failed, an error will be returned.
+func (a *Adapter) getReleasePipelineRun() (*v1beta1.PipelineRun, error) {
+	pipelineRuns := &v1beta1.PipelineRunList{}
+	opts := []client.ListOption{
+		client.MatchingLabels{
+			tekton.ReleaseNameLabel:      a.release.Name,
+			tekton.ReleaseWorkspaceLabel: a.release.Namespace,
+		},
+	}
+
+	err := a.client.List(a.context, pipelineRuns, opts...)
+	if err != nil {
+		return nil, err
+	} else if len(pipelineRuns.Items) == 0 {
+		return nil, nil
+	}
+
+	return &pipelineRuns.Items[0], nil
+}
+
+// getReleaseStrategy returns the ReleaseStrategy referenced by the given ReleaseLink. If the ReleaseLink is in the
+// same namespace as the Release being processed, it is assumed that the ReleaseLink containing the ReleaseStrategy
+// will be found in the target workspace referenced by the given ReleaseLink. If the ReleaseStrategy is
+// not found or the Get operation failed, an error will be returned.
+func (a *Adapter) getReleaseStrategy(releaseLink *v1alpha1.ReleaseLink) (*v1alpha1.ReleaseStrategy, error) {
+	// Check if the ReleaseLink is in the Release workspace. If so, we need to get the matching ReleaseLink
+	// in the target workspace.
+	if releaseLink.Namespace == a.release.Namespace {
+		var err error
+		releaseLink, err = a.getTargetReleaseLink(releaseLink)
+		if err != nil {
+			return nil, fmt.Errorf("failed to find a matching ReleaseLink in target workspace (%s)",
+				releaseLink.Spec.Target)
+		}
+	}
+
+	releaseStrategy := &v1alpha1.ReleaseStrategy{}
+	err := a.client.Get(a.context, types.NamespacedName{
+		Name:      releaseLink.Spec.ReleaseStrategy,
+		Namespace: releaseLink.Namespace,
+	}, releaseStrategy)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return releaseStrategy, nil
+}
+
+// getTargetReleaseLink returns the ReleaseLink referenced by the given ReleaseLink. If a matching ReleaseLink is not
+// found or List operation failed, an error will be returned.
+func (a *Adapter) getTargetReleaseLink(releaseLink *v1alpha1.ReleaseLink) (*v1alpha1.ReleaseLink, error) {
+	releaseLinks := &v1alpha1.ReleaseLinkList{}
+	opts := []client.ListOption{
+		client.InNamespace(releaseLink.Spec.Target),
+		client.MatchingFields{"spec.target": releaseLink.Namespace},
+	}
+
+	err := a.client.List(a.context, releaseLinks, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, foundReleaseLink := range releaseLinks.Items {
+		if foundReleaseLink.Spec.Application == releaseLink.Spec.Application {
+			return &foundReleaseLink, nil
+		}
+	}
+
+	return nil, fmt.Errorf("no ReleaseLink found in target workspace '%s' with target '%s' and application '%s'",
```

Comment:
I am not sure why we sometimes log errors with `fmt` and sometimes don't (for example see the `getReleasePipelineRun` which is almost identical)